File: Advanced/ViewPagerDemo/app/src/main/java/edu/cs4730/viewpagerdemo/FragLeft.java
package edu.cs4730.viewpagerdemo;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.fragment.app.Fragment;

public class FragLeft extends Fragment {
    TextView tx;
    String savedData;
    String TAG = "Left";

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Log.d(TAG, "OnCreate");
        if (savedInstanceState != null) {
            Log.d(TAG, "OnCreate savedInstanceState");
            savedData = savedInstanceState.getString("text");
        } else {
            savedData = "";  //just making sure.
        }

    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        Log.d(TAG, "OnActivityCreate");
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        Log.d("Left", "OnCreateView");
        View view = inflater.inflate(R.layout.left, container, false);
        tx = (TextView) view.findViewById(R.id.tvleft);
        setText(savedData);
        return view;
    }

    public void setText(String str) {
        tx.setText(tx.getText() + "\n" + str);
    }

    @Override
    public void onPause() {
        super.onPause();
        Log.i(TAG, "onPause()");

    }
    @Override
    public void onResume() {
        super.onResume();
        Log.i(TAG, "onResume()");

    }
    @Override
    public void onDestroyView() {
        super.onDestroyView();
        Log.i(TAG, "onDestroyView()");

    }
    @Override
    public void onSaveInstanceState(Bundle outState) {
        Log.d(TAG, "OnSaveInstanceState");
        super.onSaveInstanceState(outState);
        outState.putString("text", tx.getText().toString());
    }
}

File: Advanced/ViewPager2Demo_kt/app/src/main/java/edu/cs4730/viewpager2demo_kt/FragLeft.kt
package edu.cs4730.viewpager2demo_kt

import android.widget.TextView
import android.os.Bundle
import android.util.Log
import androidx.lifecycle.ViewModelProvider
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment

/**
 * This is a simple fragment to display data and it the "left" most fragment in the viewpager.
 * The code here is identical to the code in the right fragment.
 */
class FragLeft : Fragment() {
    lateinit var tx: TextView
    lateinit var mViewModel: DataViewModel
    var TAG = "Left"
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Log.d(TAG, "OnCreate")
        if (savedInstanceState != null) {
            Log.d(TAG, "OnCreate savedInstanceState")
        }
        mViewModel = ViewModelProvider(requireActivity()).get(DataViewModel::class.java)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        Log.d("Left", "OnCreateView")
        val view = inflater.inflate(R.layout.left, container, false)
        tx = view.findViewById(R.id.tvleft)
        mViewModel.dataLeft.observe(viewLifecycleOwner, { data -> tx.setText(data) })
        return view
    }

    override fun onPause() {
        super.onPause()
        Log.d(TAG, "onPause()")
    }

    override fun onResume() {
        super.onResume()
        Log.d(TAG, "onResume()")
    }

    override fun onDestroyView() {
        super.onDestroyView()
        Log.d(TAG, "onDestroyView()")
    }

    override fun onSaveInstanceState(outState: Bundle) {
        Log.d(TAG, "OnSaveInstanceState")
        super.onSaveInstanceState(outState)
    }
}